File: Java/src/net/mirchs/dm8/graphics/layers/Layer.java
package net.mirchs.dm8.graphics.layers;

import java.util.ArrayList;

import net.mirchs.dm8.graphics.Renderable2D;
import net.mirchs.dm8.graphics.Renderer2D;
import net.mirchs.dm8.graphics.Shader;
import net.mirchs.dm8.math.Matrix4f;
import net.mirchs.dm8.math.Vector3f;

public abstract class Layer {

	protected Renderer2D renderer;
	protected ArrayList<Renderable2D> renderables;
	protected Shader shader;
	protected Matrix4f pr_matrix;

	protected Layer(Renderer2D renderer, Shader shader, Matrix4f projectionMatrix) {
		this.renderer = renderer;
		this.shader = shader;
		this.pr_matrix = projectionMatrix;
		renderables = new ArrayList<>();

		shader.enable();
		shader.setUniformMat4f("pr_matrix", pr_matrix);
		shader.disable();

	}

	public Layer() {

		shader.enable();
		shader.setUniformMat4f("pr_matrix", pr_matrix);
		shader.disable();
	}

	public void add(Renderable2D renderable) {
		renderables.add(renderable);
	}

	public void render() {
		shader.enable();
		renderer.begin();
		for (Renderable2D renderable : renderables) {
			renderable.submit(renderer);
		}
		renderer.pop();
		renderer.end();
		renderer.flush();
		shader.disable();
	}

}

File: C++/DM8-CLI/src/graphics/layers/Layer.cpp
#include "Layer.h"


namespace DM8CLI {
	namespace graphics {

		Layer::Layer(DM8::graphics::Layer* instance)
			: DMManagedObject(instance)
		{
		}

		Layer::Layer(Renderer2D^ renderer, Shader^ shader, math::Mat4^ projectionMatrix)
		{
			// hard coded
			//DM8::math::Mat4 matrix = DM8::math::Mat4::Ortho(0.0f, 16.0f, 0.0f, 9.0f, -1.0f, 1.0f);
			DM8::graphics::Renderer2D* pRenderer = renderer->GetHandle();
			DM8::graphics::Shader* pShader = shader->GetHandle();
			std::cout << "cur" << std::endl;
			DM8::math::Mat4* matrixTest = projectionMatrix->GetHandle();


			m_Instance = new DM8::graphics::Layer(pRenderer, pShader, *matrixTest);
		}

		void Layer::Push(Renderable2D^ renderable)
		{
			m_Instance->Push(renderable->GetHandle());
		}

		void Layer::Render()
		{
			m_Instance->Render();
		}

	}
}